### ISSUE TITLE
`java.lang.Record` equality is much slower than HotSpot

### ISSUE THREAD

**Issue by harpocrates** [eclipse-openj9/openj9#12359]
__TL;DR: JDK16 records use `invokedynamic` for equality checks and that behaves very badly on OpenJ9 (eg. compared to manually written equality checks).__

Record classes introduced in JDK16 seem to be quite a bit slower on OpenJ9 than Hotspot. I put together a [simple JMH benchmark][0] to compare the performance of `equals` on Java records to `equals` on Scala case classes. Looking at the bytecode the interesting difference is:

  * Java records use `invokedynamic` on [`java.lang.runtime.ObjectMethods`][1]
  * Scala case classes generate bytecode that manually checks each field for equality

Here's the relevant snippets from `javap -p -v`:

<details>

<summary>manually written equality (Scala <code>case class</code>)</summary>

```
  public boolean equals(java.lang.Object);
    descriptor: (Ljava/lang/Object;)Z
    flags: ACC_PUBLIC
    Code:
      stack=4, locals=6, args_size=2
         0: aload_0
         1: aload_1
         2: if_acmpeq     102
         5: aload_1
         6: astore_3
         7: aload_3
         8: instanceof    #2                  // class records/SmallScala
        11: ifeq          19
        14: iconst_1
        15: istore_2
        16: goto          27
        19: goto          22
        22: iconst_0
        23: istore_2
        24: goto          27
        27: iload_2
        28: ifeq          106
        31: aload_1
        32: checkcast     #2                  // class records/SmallScala
        35: astore        4
        37: aload_0
        38: invokevirtual #62                 // Method i:()I
        41: aload         4
        43: invokevirtual #62                 // Method i:()I
        46: if_icmpne     98
        49: aload_0
        50: invokevirtual #65                 // Method s:()Ljava/lang/String;
        53: aload         4
        55: invokevirtual #65                 // Method s:()Ljava/lang/String;
        58: astore        5
        60: dup
        61: ifnonnull     73
        64: pop
        65: aload         5
        67: ifnull        81
        70: goto          98
        73: aload         5
        75: invokevirtual #137                // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
        78: ifeq          98
        81: aload_0
        82: invokevirtual #68                 // Method l:()J
        85: aload         4
        87: invokevirtual #68                 // Method l:()J
        90: lcmp
        91: ifne          98
        94: iconst_1
        95: goto          99
        98: iconst_0
        99: ifeq          106
       102: iconst_1
       103: goto          107
       106: iconst_0
       107: ireturn
```

</details>

<details>
<summary><code>invokedynamic</code> equality (Java <code>record</code>)</summary>

```
  public final boolean equals(java.lang.Object);
    descriptor: (Ljava/lang/Object;)Z
    flags: (0x0011) ACC_PUBLIC, ACC_FINAL
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokedynamic #29,  0             // InvokeDynamic #0:equals:(Lrecords/SmallJava;Ljava/lang/Object;)Z
         7: ireturn
...
BootstrapMethods:
  0: #45 REF_invokeStatic java/lang/runtime/ObjectMethods.bootstrap:(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/TypeDescriptor;Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/invoke/MethodHandle;)Ljava/lang/Object;
    Method arguments:
      #8 records/SmallJava
      #52 i;s;l
      #54 REF_getField records/SmallJava.i:I
      #55 REF_getField records/SmallJava.s:Ljava/lang/String;
      #56 REF_getField records/SmallJava.l:J
```

</details>

I ran this benchmark twice to compare HotSpot OpenJDK 16 and Eclipse OpenJ9. Although disappointing, it isn't too surprising that OpenJ9 isn't faster (after all, I don't think this benchmark exercises any of OpenJ9's strengths). What _is_ surprising is that `invokedynamic` on OpenJ9 is abysmal and `invokedynamic` on HotSpot is the best.

![Unknown-5](https://user-images.githubusercontent.com/10766081/113728899-5a179100-96ab-11eb-8fcf-56010a494b5b.png)

|                         | HotSpot "vanilla" | HotSpot `invokedynamic` | OpenJ9 "vanilla" | OpenJ9 `invokedynamic` |
| ----------------------- | ----------------- | ----------------------- | ---------------- | ---------------------- |
| Large record equality   | 32.445 ± 7.299    |     9.588 ± 0.097       | 58.364 ± 6.539   |  133.907 ± 3.924       |
| Large record inequality |  6.622 ± 0.071    |     3.418 ± 0.592       |  6.710 ± 1.031   |   31.420 ± 1.408       |
| Small record equality   |  7.083 ± 0.419    |     3.892 ± 0.470       | 12.112 ± 1.443   |   32.535 ± 0.341       |
| Small record inequality |  3.294 ± 0.131    |     2.847 ± 0.061       |  2.808 ± 1.137   |   44.639 ± 6.619       |

---

OpenJ9 (installed with `brew install  adoptopenjdk16-openj9`)

```
$ java -version
openjdk version "16" 2021-03-16
OpenJDK Runtime Environment AdoptOpenJDK (build 16+36)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.25.0, JRE 16 Mac OS X amd64-64-Bit Compressed References 20210311_66 (JIT enabled, AOT enabled)
OpenJ9   - 022d65424
OMR      - 09514431e
JCL      - 0c11227a21a based on jdk-16+36)
```

HotSpot OpenJDK 16 (installed with ` brew install adoptopenjdk16`)

```
$ java -version
openjdk version "16" 2021-03-16
OpenJDK Runtime Environment AdoptOpenJDK (build 16+36)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 16+36, mixed mode, sharing)
```

  [0]: https://github.com/harpocrates/scala-java-records-benchmark
  [1]: https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/runtime/ObjectMethods.html

**Comment by DanHeidinga** [comment]
@harpocrates there is work going on to transition from OpenJ9's current MethodHandle implementation to OpenJDK's LambdaForm implementation.  See the project tracking it: [1]

There are a number of issues with our current implementation that result in the current performance potholes.  The transition is currently targeted for the 0.27.0 OpenJ9 release (July quarterly update).

fyi @babsingh @0xdaryl 

[1] https://github.com/eclipse/openj9/projects/18

**Comment by harpocrates** [comment]
Thanks for the quick response! I'll follow that project and check the next release.

**Comment by babsingh** [comment]
I just tried to run the **simple JMH benchmark** (https://github.com/harpocrates/scala-java-records-benchmark) with OpenJ9 JDK16 + OJDK MHs (https://github.com/eclipse/openj9/projects/18).

### Errors were seen at `sbt` startup:
<details>
  <summary>Click here for the full stacktrace</summary>
  
```
Exception in thread "sbt-parser-init-thread" java.lang.ExceptionInInitializerError
	at java.base/java.lang.J9VMInternals.ensureError(J9VMInternals.java:184)
	at java.base/java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:173)
	at sbt.internal.parser.SbtParserInit$$anon$2.run(SbtParser.scala:171)
Caused by: java.lang.IllegalStateException: uninitialized call site
	at java.base/java.lang.invoke.CallSite.uninitializedCallSite(CallSite.java:261)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x0000000000000000.invokeStatic(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invokeExact_MT(LambdaForm$MH)
	at scala.reflect.internal.util.StatisticsStatics.areSomeColdStatsEnabled(StatisticsStatics.java:43)
	at scala.reflect.internal.Symbols$TypeSymbol.<init>(Symbols.scala:3223)
	at scala.reflect.internal.Symbols$ClassSymbol.<init>(Symbols.scala:3279)
	at scala.reflect.internal.Symbols$ModuleClassSymbol.<init>(Symbols.scala:3444)
	at scala.reflect.internal.Symbols$PackageClassSymbol.<init>(Symbols.scala:3493)
	at scala.reflect.internal.Mirrors$Roots$RootClass.<init>(Mirrors.scala:308)
	at scala.reflect.internal.Mirrors$Roots.RootClass$lzycompute(Mirrors.scala:323)
	at scala.reflect.internal.Mirrors$Roots.RootClass(Mirrors.scala:323)
	at scala.reflect.internal.Mirrors$Roots$EmptyPackageClass.<init>(Mirrors.scala:332)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass$lzycompute(Mirrors.scala:338)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:338)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:278)
	at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:252)
	at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:75)
	at scala.tools.nsc.Global.rootMirror(Global.scala:73)
	at scala.tools.nsc.Global.rootMirror(Global.scala:45)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:294)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:294)
	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1503)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1207)
	at sbt.internal.parser.SbtParser$.<init>(SbtParser.scala:121)
	at sbt.internal.parser.SbtParser$.<clinit>(SbtParser.scala)
	... 1 more
[info] welcome to sbt 1.4.9 (Eclipse OpenJ9 Java 16-internal)
java.lang.NoClassDefFoundError: sbt.internal.parser.SbtParser$ (initialization failure)
	at java.base/java.lang.J9VMInternals.initializationAlreadyFailed(J9VMInternals.java:134)
	at sbt.internal.parser.SbtParser.splitExpressions(SbtParser.scala:227)
	at sbt.internal.parser.SbtParser.<init>(SbtParser.scala:216)
	at sbt.internal.EvaluateConfigurations$.splitExpressions(EvaluateConfigurations.scala:288)
	at sbt.internal.EvaluateConfigurations$.parseConfiguration(EvaluateConfigurations.scala:97)
	at sbt.internal.EvaluateConfigurations$.evaluateSbtFile(EvaluateConfigurations.scala:146)
	at sbt.internal.Load$.loadSettingsFile$1(Load.scala:1106)
	at sbt.internal.Load$.$anonfun$discoverProjects$2(Load.scala:1116)
	at sbt.internal.Load$$$Lambda$1464/0x000000004d6722c8.apply(Unknown Source)
	at scala.collection.MapLike.getOrElse(MapLike.scala:131)
	at scala.collection.MapLike.getOrElse$(MapLike.scala:129)
	at scala.collection.AbstractMap.getOrElse(Map.scala:65)
	at sbt.internal.Load$.memoLoadSettingsFile$1(Load.scala:1115)
	at sbt.internal.Load$.$anonfun$discoverProjects$4(Load.scala:1123)
	at sbt.internal.Load$$$Lambda$1463/0x000000004d671e78.apply(Unknown Source)
	at scala.collection.immutable.List.map(List.scala:293)
	at sbt.internal.Load$.loadFiles$1(Load.scala:1123)
	at sbt.internal.Load$.discoverProjects(Load.scala:1137)
	at sbt.internal.Load$.discover$1(Load.scala:889)
	at sbt.internal.Load$.loadTransitive(Load.scala:943)
	at sbt.internal.Load$.loadProjects$1(Load.scala:726)
	at sbt.internal.Load$.$anonfun$loadUnit$11(Load.scala:729)
	at sbt.internal.Load$$$Lambda$1456/0x000000004d65ce50.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.$anonfun$loadUnit$1(Load.scala:729)
	at sbt.internal.Load$$$Lambda$1424/0x000000004d2d1c40.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.loadUnit(Load.scala:688)
	at sbt.internal.Load$.$anonfun$builtinLoader$4(Load.scala:486)
	at sbt.internal.Load$$$Lambda$1420/0x000000004d2d1058.apply(Unknown Source)
	at sbt.internal.BuildLoader$.$anonfun$componentLoader$5(BuildLoader.scala:180)
	at sbt.internal.BuildLoader$$$Lambda$1422/0x000000004d2d1640.apply(Unknown Source)
	at sbt.internal.BuildLoader.apply(BuildLoader.scala:245)
	at sbt.internal.Load$.loadURI$1(Load.scala:548)
	at sbt.internal.Load$.loadAll(Load.scala:564)
	at sbt.internal.Load$.loadURI(Load.scala:494)
	at sbt.internal.Load$.load(Load.scala:473)
	at sbt.internal.Load$.$anonfun$apply$1(Load.scala:242)
	at sbt.internal.Load$$$Lambda$1394/0x000000004d2b9858.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.apply(Load.scala:242)
	at sbt.internal.Load$.buildPluginDefinition(Load.scala:1299)
	at sbt.internal.Load$.buildPlugins(Load.scala:1229)
	at sbt.internal.Load$.plugins(Load.scala:1208)
	at sbt.internal.Load$.$anonfun$loadUnit$2(Load.scala:694)
	at sbt.internal.Load$$$Lambda$1425/0x000000004d2d2738.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.$anonfun$loadUnit$1(Load.scala:694)
	at sbt.internal.Load$$$Lambda$1424/0x000000004d2d1c40.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.loadUnit(Load.scala:688)
	at sbt.internal.Load$.$anonfun$builtinLoader$4(Load.scala:486)
	at sbt.internal.Load$$$Lambda$1420/0x000000004d2d1058.apply(Unknown Source)
	at sbt.internal.BuildLoader$.$anonfun$componentLoader$5(BuildLoader.scala:180)
	at sbt.internal.BuildLoader$$$Lambda$1422/0x000000004d2d1640.apply(Unknown Source)
	at sbt.internal.BuildLoader.apply(BuildLoader.scala:245)
	at sbt.internal.Load$.loadURI$1(Load.scala:548)
	at sbt.internal.Load$.loadAll(Load.scala:564)
	at sbt.internal.Load$.loadURI(Load.scala:494)
	at sbt.internal.Load$.load(Load.scala:473)
	at sbt.internal.Load$.$anonfun$apply$1(Load.scala:242)
	at sbt.internal.Load$$$Lambda$1394/0x000000004d2b9858.apply(Unknown Source)
	at sbt.internal.Load$.timed(Load.scala:1382)
	at sbt.internal.Load$.apply(Load.scala:242)
	at sbt.internal.Load$.defaultLoad(Load.scala:55)
	at sbt.BuiltinCommands$.liftedTree1$1(Main.scala:934)
	at sbt.BuiltinCommands$.doLoadProject(Main.scala:934)
	at sbt.BuiltinCommands$.$anonfun$loadProjectImpl$2(Main.scala:887)
	at sbt.BuiltinCommands$$$Lambda$993/0x00000000b01aae80.apply(Unknown Source)
	at sbt.Command$.$anonfun$applyEffect$4(Command.scala:150)
	at sbt.Command$$$Lambda$835/0x000000004d061b58.apply(Unknown Source)
	at sbt.Command$.$anonfun$applyEffect$2(Command.scala:145)
	at sbt.Command$$$Lambda$905/0x00000000b01173e0.apply(Unknown Source)
	at sbt.Command$.process(Command.scala:189)
	at sbt.MainLoop$.$anonfun$processCommand$5(MainLoop.scala:243)
	at sbt.MainLoop$$$Lambda$815/0x000000004d05ea18.apply(Unknown Source)
	at scala.Option.getOrElse(Option.scala:189)
	at sbt.MainLoop$.process$1(MainLoop.scala:243)
	at sbt.MainLoop$.processCommand(MainLoop.scala:276)
	at sbt.MainLoop$.$anonfun$next$5(MainLoop.scala:163)
	at sbt.MainLoop$$$Lambda$758/0x00000000b00ee890.apply(Unknown Source)
	at sbt.State$StateOpsImpl$.runCmd$1(State.scala:289)
	at sbt.State$StateOpsImpl$.process$extension(State.scala:325)
	at sbt.MainLoop$.$anonfun$next$4(MainLoop.scala:163)
	at sbt.MainLoop$$$Lambda$757/0x00000000b00ee128.apply(Unknown Source)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
	at sbt.MainLoop$.next(MainLoop.scala:163)
	at sbt.MainLoop$.run(MainLoop.scala:144)
	at sbt.MainLoop$.$anonfun$runWithNewLog$1(MainLoop.scala:119)
	at sbt.MainLoop$$$Lambda$702/0x000000004d0027d8.apply(Unknown Source)
	at sbt.io.Using.apply(Using.scala:27)
	at sbt.MainLoop$.runWithNewLog(MainLoop.scala:112)
	at sbt.MainLoop$.runAndClearLast(MainLoop.scala:66)
	at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:51)
	at sbt.MainLoop$.runLogged(MainLoop.scala:42)
	at sbt.StandardMain$.runManaged(Main.scala:193)
	at sbt.xMain$.$anonfun$run$8(Main.scala:102)
	at sbt.xMain$$$Lambda$241/0x00000000bc352690.apply(Unknown Source)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withIn(Console.scala:230)
	at sbt.internal.util.Terminal$.withIn(Terminal.scala:560)
	at sbt.internal.util.Terminal$.$anonfun$withStreams$1(Terminal.scala:350)
	at sbt.internal.util.Terminal$$$Lambda$355/0x00000000b80c9f08.apply(Unknown Source)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withOut(Console.scala:167)
	at sbt.internal.util.Terminal$.$anonfun$withOut$2(Terminal.scala:550)
	at sbt.internal.util.Terminal$$$Lambda$359/0x00000000b00cfb10.apply(Unknown Source)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withErr(Console.scala:196)
	at sbt.internal.util.Terminal$.withOut(Terminal.scala:550)
	at sbt.internal.util.Terminal$.withStreams(Terminal.scala:350)
	at sbt.xMain$.run(Main.scala:86)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.scala:83)
	at sbt.xMain.run(Main.scala:46)
	at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:111)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:131)
	at xsbt.boot.Launch$.run(Launch.scala:111)
	at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:37)
	at xsbt.boot.Launch$.launch(Launch.scala:120)
	at xsbt.boot.Launch$.apply(Launch.scala:20)
	at xsbt.boot.Boot$.runImpl(Boot.scala:56)
	at xsbt.boot.Boot$.main(Boot.scala:18)
	at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.lang.IllegalStateException: uninitialized call site
	at java.base/java.lang.invoke.CallSite.uninitializedCallSite(CallSite.java:261)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x0000000000000000.invokeStatic(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000000000000.invokeExact_MT(LambdaForm$MH)
	at scala.reflect.internal.util.StatisticsStatics.areSomeColdStatsEnabled(StatisticsStatics.java:43)
	at scala.reflect.internal.Symbols$TypeSymbol.<init>(Symbols.scala:3223)
	at scala.reflect.internal.Symbols$ClassSymbol.<init>(Symbols.scala:3279)
	at scala.reflect.internal.Symbols$ModuleClassSymbol.<init>(Symbols.scala:3444)
	at scala.reflect.internal.Symbols$PackageClassSymbol.<init>(Symbols.scala:3493)
	at scala.reflect.internal.Mirrors$Roots$RootClass.<init>(Mirrors.scala:308)
	at scala.reflect.internal.Mirrors$Roots.RootClass$lzycompute(Mirrors.scala:323)
	at scala.reflect.internal.Mirrors$Roots.RootClass(Mirrors.scala:323)
	at scala.reflect.internal.Mirrors$Roots$EmptyPackageClass.<init>(Mirrors.scala:332)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass$lzycompute(Mirrors.scala:338)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:338)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:278)
	at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:252)
	at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:75)
	at scala.tools.nsc.Global.rootMirror(Global.scala:73)
	at scala.tools.nsc.Global.rootMirror(Global.scala:45)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:294)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:294)
	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1503)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1207)
	at sbt.internal.parser.SbtParser$.<init>(SbtParser.scala:121)
	at sbt.internal.parser.SbtParser$.<clinit>(SbtParser.scala)
	at sbt.internal.parser.SbtParserInit$$anon$2.run(SbtParser.scala:171)
```
</details

fyi @fengxue-IS

**Comment by harpocrates** [comment]
Let me know if you want me to translate this to a simple vanilla JMH project. I was using SBT JMH because I'm investigating changing the generated code in Scala case classes, but I could try to cobble together a version of this benchmark that is just vanilla Java and JMH.

**Comment by babsingh** [comment]
> try to cobble together a version of this benchmark that is just vanilla Java and JMH

Currently, we are resolving a number of functional deficiencies for OJDK MHs in OpenJ9. Once https://github.com/eclipse/openj9/projects/18 is functionally stable, we should be able to run the benchmark in the SBT JMH framework. So, there is no need to convert the benchmark. I will post perf numbers once OpenJ9 minimally supports OJDK MHs in order to run the benchmark.

**Comment by babsingh** [comment]
https://github.com/eclipse/openj9/issues/12404 has been opened to address the **`sbt`** startup error reported in https://github.com/eclipse/openj9/issues/12359#issuecomment-814406590.

**Comment by babsingh** [comment]
### Results: -Xint (no JIT)
**Summary:**
- OpenJDK16 has the slowest interpreter performance.
- OpenJ9 JDK16 has the best interpreter performance.
- "OpenJ9 + OJDK-MH" JDK16 is 2x faster than OpenJDK16 and 2x slower than OpenJ9 JDK16.

**OpenJ9 JDK16:**
```
[info] Benchmark                       Mode  Cnt     Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  9950.584 ± 188.170  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250  1164.408 ±  21.880  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250  1885.655 ±  33.527  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250   676.182 ±  11.971  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250  1954.278 ±  36.958  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250  1152.360 ±  22.672  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250   401.999 ±   7.901  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250   306.968 ±   5.983  ns/op
[success] Total time: 1119 s (18:39), completed Apr 13, 2021, 9:50:58 PM
```

**OpenJDK16:**
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  50405.302 ± 887.352  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   4730.219 ±  85.907  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250   4906.817 ±  87.256  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250   1871.964 ±  35.850  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   9344.070 ± 163.555  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   4656.625 ±  87.082  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250   1040.651 ±  19.588  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250    790.994 ±  13.875  ns/op
[success] Total time: 1015 s (16:55), completed Apr 13, 2021, 10:24:01 PM
```

**"OpenJ9 + OJDK-MH" JDK16:**
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  23365.353 ± 473.677  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   2451.106 ±  53.738  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250   1875.820 ±  33.689  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250    673.542 ±  13.545  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   4514.004 ±  97.779  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   2444.818 ±  47.399  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250    404.062 ±   8.365  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250    315.198 ±   6.099  ns/op
[success] Total time: 1151 s (19:11), completed Apr 13, 2021, 11:00:38 PM
```

**Comment by babsingh** [comment]
### Perf Results: -Xint (no JIT)
**Summary:**
- **OpenJDK16** has the slowest interpreter performance.
- **OpenJ9 JDK16** has the best interpreter performance.
- **"OpenJ9 + OJDK-MH" JDK16** is 2x faster than **OpenJDK16** and 2x slower than **OpenJ9 JDK16**.

**OpenJ9 JDK16:**
```
[info] Benchmark                       Mode  Cnt     Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  9950.584 ± 188.170  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250  1164.408 ±  21.880  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250  1885.655 ±  33.527  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250   676.182 ±  11.971  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250  1954.278 ±  36.958  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250  1152.360 ±  22.672  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250   401.999 ±   7.901  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250   306.968 ±   5.983  ns/op
[success] Total time: 1119 s (18:39), completed Apr 12, 2021, 9:50:58 PM
```

**OpenJDK16:**
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  50405.302 ± 887.352  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   4730.219 ±  85.907  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250   4906.817 ±  87.256  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250   1871.964 ±  35.850  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   9344.070 ± 163.555  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   4656.625 ±  87.082  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250   1040.651 ±  19.588  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250    790.994 ±  13.875  ns/op
[success] Total time: 1015 s (16:55), completed Apr 12, 2021, 10:24:01 PM
```

**"OpenJ9 + OJDK-MH" JDK16:**
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  23365.353 ± 473.677  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   2451.106 ±  53.738  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250   1875.820 ±  33.689  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250    673.542 ±  13.545  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   4514.004 ±  97.779  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   2444.818 ±  47.399  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250    404.062 ±   8.365  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250    315.198 ±   6.099  ns/op
[success] Total time: 1151 s (19:11), completed Apr 12, 2021, 11:00:38 PM
```

### Perf Results: JIT enabled (no -Xint)

**Summary:**
- **"OpenJ9 + OJDK-MH" JDK16** has the slowest performance. It is in the same order of magnitude as its `-Xint` performance. JIT optimizations are still being implemented. @liqunl for confirmation?
- **OpenJDK16** has the best performance.
- **OpenJ9 JDK16** performance is somewhere in the middle.

**OpenJ9 JDK16:**
```
[info] Benchmark                       Mode  Cnt    Score   Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  192.941 ± 5.353  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   45.023 ± 1.755  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250   48.916 ± 4.290  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250    9.812 ± 0.247  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   44.770 ± 1.000  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   37.989 ± 0.848  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250   16.286 ± 1.360  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250    3.676 ± 0.326  ns/op
[success] Total time: 994 s (16:34), completed Apr 12, 2021, 9:31:59 PM
```

**OpenJDK16:**
```
[info] Benchmark                       Mode  Cnt   Score   Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  14.988 ± 0.276  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   5.059 ± 0.098  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250  48.414 ± 0.944  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250   9.935 ± 0.186  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   5.648 ± 0.116  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   5.281 ± 0.113  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250  10.611 ± 0.399  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250   4.826 ± 0.097  ns/op
[success] Total time: 943 s (15:43), completed Apr 12, 2021, 10:06:47 PM
```

**"OpenJ9 + OJDK-MH" JDK16:**
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] Bench.largeJavaEqualityTest     avgt  250  18429.867 ± 365.013  ns/op
[info] Bench.largeJavaInequalityTest   avgt  250   1894.562 ±  47.936  ns/op
[info] Bench.largeScalaEqualityTest    avgt  250     48.489 ±   4.616  ns/op
[info] Bench.largeScalaInequalityTest  avgt  250      9.866 ±   0.216  ns/op
[info] Bench.smallJavaEqualityTest     avgt  250   3477.044 ±  86.548  ns/op
[info] Bench.smallJavaInequalityTest   avgt  250   1893.541 ±  51.372  ns/op
[info] Bench.smallScalaEqualityTest    avgt  250     16.785 ±   1.237  ns/op
[info] Bench.smallScalaInequalityTest  avgt  250      3.770 ±   0.331  ns/op
[success] Total time: 1016 s (16:56), completed Apr 12, 2021, 10:41:07 PM
```

fyi @nbhuiyan @fengxue-IS @jdmpapin @tajila

**Comment by liqunl** [comment]
We're still working on the performance of **"OpenJ9 + OJDK-MH" JDK16**, the gaps to **OpenJ9 JDK16** and **OpenJDK16** are expected. The gap between **OpenJ9 JDK16** and **OpenJDK16**  is also a known issue, we've seen it on jruby and nashorn, but we're not looking into it as we're switching to **"OpenJ9 + OJDK-MH" JDK16**

**Comment by pshipton** [comment]
Moving forward.
